File: Medium Problem's/p-155. Min Stack/Solution.js
//Satyam-R4j-LeetCode Problem-155 (medium)
//Solution in JavaScript
var MinStack = function ()
{

    this.stack = []
    this.minStack = []
};

MinStack.prototype.push = function (val)
{
    this.stack.push(val)
    if (this.minStack.length === 0 || val <= this.minStack[this.minStack.length - 1])
    {
        this.minStack.push(val)
    }
};

MinStack.prototype.pop = function ()
{

    const val = this.stack.pop()
    if (val === this.minStack[this.minStack.length - 1])
    {
        this.minStack.pop()
    }

};

MinStack.prototype.top = function ()
{
    return this.stack[this.stack.length - 1]
};

MinStack.prototype.getMin = function ()
{

    return this.minStack[this.minStack.length - 1]
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

const minStack = new MinStack()
minStack.push(-1)
minStack.push(0)
minStack.push(-3)

console.log(minStack.getMin());
minStack.pop()
console.log(minStack.getMin());
console.log(minStack.top());

File: Medium Problem's/p-155. Min Stack/Solution.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.minVal = None

    def push(self, val: int) -> None:
        if not self.stack:
            self.stack.append(val)
            self.minVal = val
        elif val >= self.minVal: # type: ignore
            self.stack.append(val)
        else:
            self.stack.append(2 * val - self.minVal) # type: ignore
            self.minVal = val

    def pop(self) -> None:
        if not self.stack:
            return
        
        if self.stack[-1] < self.minVal:
            self.minVal = 2 * self.minVal - self.stack[-1] # type: ignore
        
        self.stack.pop()
        
        if not self.stack:
            self.minVal = None

    def top(self) -> int:
        if not self.stack:
            return None # type: ignore
        if self.stack[-1] >= self.minVal:
            return self.stack[-1]
        return self.minVal # type: ignore

    def getMin(self) -> int:
        if not self.stack:
            return None # type: ignore
        return self.minVal # type: ignore